Popover light dismiss does not work when touch event regions are enabled

lilyspiniolas opened the pull request:
#### c467303c1538dce11f24ec1c91d7d8ed503d6479
<pre>
Popover light dismiss does not work when touch event regions are enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=303527">https://bugs.webkit.org/show_bug.cgi?id=303527</a>
<a href="https://rdar.apple.com/165823080">rdar://165823080</a>

Reviewed by Tim Nguyen and Abrar Rahman Protyasha.

A pointerup event followed by a pointerdown event occuring outside of a
popover should cause the popover to dismiss. This functionality is broken with
touch event regions enabled because touch event regions are only created for
elements with JS event listeners or for elements with internal touch handling.
Neither of these cases apply when a popover is shown.

To resolve this issue, call `Document::invalidateEventListenerRegions()` when a
top-layer element (popover) is created or removed if the value of
`needsPointerEventHandlingForPopover()` changed. This results in
`Style::Adjuster::adjustEventListenerRegionTypesForRootStyle` being called, where
we now add `EventListenerRegionTypes::PointerDown` &amp;&amp; `PointerUp` to the root
style if the document needs pointer event handling for a popover. As a result,
the missing event region is now generated.

Tested by
imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-with-anchor.tentative.html

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addTopLayerElement):
(WebCore::Document::removeTopLayerElement):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustEventListenerRegionTypesForRootStyle):

Canonical link: <a href="https://commits.webkit.org/303923@main">https://commits.webkit.org/303923@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f5bb733bb37c4f9346e1dda8fd2aa6a2a300964

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141516 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85997 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9d55f98f-e363-4644-9fde-cc7a937986cc) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135806 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6978 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6311 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102450 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136883 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83249 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4783 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2402 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113944 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38211 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144161 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6117 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38790 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110821 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6199 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5193 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111029 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28172 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4638 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116349 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59864 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6169 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34602 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6015 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69633 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6260 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6123 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
Seems fine

aprotyas reviewed (approved):
> Tested by
imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-with-anchor.tentative.html

Should we expect a diff in the expectations? (at least, in configs where the test isn't skipped)

lilyspiniolas committed:
Popover light dismiss does not work when touch event regions are enabled
https://bugs.webkit.org/show_bug.cgi?id=303527
rdar://165823080

Reviewed by Tim Nguyen and Abrar Rahman Protyasha.

A pointerup event followed by a pointerdown event occuring outside of a
popover should cause the popover to dismiss. This functionality is broken with
touch event regions enabled because touch event regions are only created for
elements with JS event listeners or for elements with internal touch handling.
Neither of these cases apply when a popover is shown.

To resolve this issue, call `Document::invalidateEventListenerRegions()` when a
top-layer element (popover) is created or removed if the value of
`needsPointerEventHandlingForPopover()` changed. This results in
`Style::Adjuster::adjustEventListenerRegionTypesForRootStyle` being called, where
we now add `EventListenerRegionTypes::PointerDown` && `PointerUp` to the root
style if the document needs pointer event handling for a popover. As a result,
the missing event region is now generated.

Tested by
imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-with-anchor.tentative.html

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addTopLayerElement):
(WebCore::Document::removeTopLayerElement):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustEventListenerRegionTypesForRootStyle):

Canonical link: https://commits.webkit.org/303923@main
